data/Dockerfiles/acme/acme.sh

```diff
@@ -56,97 +46,21 @@ mkdir -p ${ACME_BASE}/acme
-  # Reloading
-  echo "Reloading Nginx..."
-  NGINX_RELOAD_RET=$(curl -X POST --insecure https://dockerapi/containers/${NGINX}/exec -d '{"cmd":"reload", "task":"nginx"}' --silent -H 'Content-type: application/json' | jq -r .type)
-  [[ ${NGINX_RELOAD_RET} != 'success' ]] && { echo "Could not reload Nginx, restarting container..."; restart_container ${NGINX} ; }
-  echo "Reloading Dovecot..."
-  DOVECOT_RELOAD_RET=$(curl -X POST --insecure https://dockerapi/containers/${DOVECOT}/exec -d '{"cmd":"reload", "task":"dovecot"}' --silent -H 'Content-type: application/json' | jq -r .type)
-  [[ ${DOVECOT_RELOAD_RET} != 'success' ]] && { echo "Could not reload Dovecot, restarting container..."; restart_container ${DOVECOT} ; }
-  echo "Reloading Postfix..."
-  POSTFIX_RELOAD_RET=$(curl -X POST --insecure https://dockerapi/containers/${POSTFIX}/exec -d '{"cmd":"reload", "task":"postfix"}' --silent -H 'Content-type: application/json' | jq -r .type)
-  [[ ${POSTFIX_RELOAD_RET} != 'success' ]] && { echo "Could not reload Postfix, restarting container..."; restart_container ${POSTFIX} ; }
-}
-
-restart_container(){
-  for container in $*; do
-    log_f "Restarting ${container}..." no_nl
-    C_REST_OUT=$(curl -X POST --insecure https://dockerapi/containers/${container}/restart | jq -r '.msg')
-    log_f "${C_REST_OUT}" no_date
-  done
-}
-
-array_diff() {
-  # https://stackoverflow.com/questions/2312762, Alex Offshore
-  eval local ARR1=\(\"\${$2[@]}\"\)
-  eval local ARR2=\(\"\${$3[@]}\"\)
-  local IFS=$'\n'
-  mapfile -t $1 < <(comm -23 <(echo "${ARR1[*]}" | sort) <(echo "${ARR2[*]}" | sort))
-}
-
-verify_hash_match(){
-  CERT_HASH=$(openssl x509 -noout -modulus -in "${1}" | openssl md5)
-  KEY_HASH=$(openssl rsa -noout -modulus -in "${2}" | openssl md5)
-  if [[ ${CERT_HASH} != ${KEY_HASH} ]]; then
-    log_f "Certificate and key hashes do not match!"
-    return 1
-  else
-    log_f "Verified hashes."
-    return 0
+if [[ -f ${ACME_BASE}/acme/key.pem && -f ${ACME_BASE}/acme/cert.pem ]]; then
+  if verify_hash_match ${ACME_BASE}/acme/cert.pem ${ACME_BASE}/acme/key.pem; then
+    log_f "Migrating to SNI folder structure..." no_nl
+    CERT_DOMAIN=($(openssl x509 -noout -text -in ${ACME_BASE}/acme/cert.pem | grep "Subject:" | sed -e 's/\(Subject:\)\|\(CN = \)\|\(CN=\)//g' | sed -e 's/^[[:space:]]*//'))
+    CERT_DOMAINS=(${CERT_DOMAIN} $(openssl x509 -noout -text -in ${ACME_BASE}/acme/cert.pem | grep "DNS:" | sed -e 's/\(DNS:\)\|,//g' | sed "s/${CERT_DOMAIN}//" | sed -e 's/^[[:space:]]*//'))
+    mkdir -p ${ACME_BASE}/${CERT_DOMAIN}
+    mv ${ACME_BASE}/acme/cert.pem ${ACME_BASE}/${CERT_DOMAIN}/cert.pem
+    # key is only copied, not moved, because it is used by all other requests too
+    cp ${ACME_BASE}/acme/key.pem ${ACME_BASE}/${CERT_DOMAIN}/key.pem
+    chmod 600 ${ACME_BASE}/${CERT_DOMAIN}/key.pem
+    echo -n ${CERT_DOMAINS[*]} > ${ACME_BASE}/${CERT_DOMAIN}/domains
```

Comment:
This is where things are going wrong. Please filter out any domains starting with _autoconfig._ or _autodiscover._. Once that's done, we can go ahead and merge this pull request.

Reply:
What if someone has configured "autoconfig.something.tld" as a SAN and "something.tld" is not setup inside mailcow? Then that domain would be missing from the certificate. Maybe we should filter all domains part of the default certificates from all additional certificates instead?